lib/factiva/authentication.rb

```diff
@@ -0,0 +1,105 @@
+require "http"
+require "json"
+
+module Factiva
+  class RequestError < Exception; end
+
+  class Authentication
+    def token
+      authz["access_token"]
+    end
+
+  private
+    def authn
+      @authn ||= set_authn
+    end
+
+    def authz
+      @authz ||= set_authz
+      @authz = set_authz(refresh: true) if expired?(@authz["expiration_timestamp"])
+      @authz
+    end
+
+    def set_authn
+      make_request(set_authn_params)
+    end
+
+    def set_authz(refresh: false)
+      params = refresh ? refresh_authz_params : set_authz_params
+
+      new_authz = make_request(params)
+      new_authz["expiration_timestamp"] = Time.now + new_authz["expires_in"]
+      new_authz
+    end
+
+    def make_request(params)
+      response = HTTP.post(
+        url,
+        params
+      )
+
+      response_body = JSON.parse(response.body.to_s)
```

Comment:
yeap, probably we should. what do you think about this instead of line 41?

```
response_body = begin
  JSON.parse(response.body.to_s)
 rescue JSON::ParserError => error
   "Failed authenticating to Factiva: #{error}"
 end
 ```
 The only problem I see is that we are assuming that a success response will be parsed correctly